### PR TITLE
Initial add of electrum 2.x mnemonic support

### PR DESCRIPTION
**Pull request by thecodefactory**

### CODE REVIEW

**Comment by thecodefactory** [comment]
Depends on libbitcoin/libbitcoin#846

Opening for review, but needs test cases (none included yet).

File: src/config/electrum_seed_type.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <bitcoin/explorer/config/electrum_seed_type.hpp>
```
Comment (evoskuil) [inline src/config/electrum_seed_type.cpp]: Style: try to avoid naming types with "type". By convention the class names for the config types match the data types names in `model/generate.xml`. In this case it should be simply "electrum".

File: model/generate.xml [hunk 2]
```diff
@@ -420,6 +420,19 @@
     <define name="BX_EC_MNEMONIC_TO_SEED_INVALID_IN_LANGUAGES" value="WARNING: The specified words are not a valid mnemonic in any supported dictionary." />
   </command>
 
+  <command symbol="electrum-new" output="string" multipleX="true" category="WALLET" description="Create a mnemonic seed (Electrum) from entropy. WARNING: mnemonic should be created from properly generated entropy.">
+    <option name="seed_type" type="electrum_seed_type" description="The seed type identifier to use. Options are 'standard', 'segwit', and 'two_factor_authentication', defaults to 'standard'." />
```
Comment (evoskuil) [inline model/generate.xml]: use:
```xml
type="electrum"
```

**Comment by thecodefactory** [comment]
Updated.

File: test/commands/electrum-to-seed.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    BX_REQUIRE_OKAY(command.invoke(output, error));
+    BX_REQUIRE_OUTPUT("c4033901dd61ba26cfd0a1cf1ceb4b347606635aa3cb951eb6e819d58beedc04dd400a2e600d783c83c75879d6538abeecc7bb1b292b2a4d775d348d5d686427\n");
+}
+
+BOOST_AUTO_TEST_CASE(electrum_to_seed__invoke__dictionary_prefix_okay_output)
+{
+    BX_DECLARE_COMMAND(electrum_to_seed);
+    command.set_language_option({ "en" });
+    command.set_prefix_option({ "standard" });
+    command.set_words_argument({ "foo", "bar", "baz" });
```
Comment (evoskuil) [inline test/commands/electrum-to-seed.cpp]: How does this work, non-dictionary words? It would be good to at least provide symmetrical tests for the seed-new tests (to ensure they round trip). We also use the tests to validate the wiki documentation.

File: test/commands/electrum-new.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    command.set_language_option({ "en" });
+    command.set_prefix_option({ "standard" });
+    BX_REQUIRE_OKAY(command.invoke(output, error));
+    BX_REQUIRE_OUTPUT("giggle crush argue inflict wear defy combine evolve tiger spatial crumble fury\n");
+}
+
+BOOST_AUTO_TEST_CASE(electrum_new__invoke__large_entropy_okay_output)
+{
+    BX_DECLARE_COMMAND(electrum_new);
+    command.set_seed_argument({ "B0756302179E800B182514C729F1D6814C377FF06097569EF540E6C1F1950F08" });
```
Comment (evoskuil) [inline test/commands/electrum-new.cpp]: Style: prefer lower case on base16 text.

File: model/generate.xml [hunk 5]
```diff
@@ -420,6 +420,20 @@
...
+    <define name="BX_EC_ELECTRUM_NEW_UNSUPPORTED" value="The electrum-new command requires an ICU build." />
+  </command>
+
+  <command symbol="electrum-to-seed" output="base16" category="WALLET" description="Convert a mnemonic seed (Electrum) to its numeric representation.">
+    <option name="language" type="language" description="The language identifier of the dictionary of the mnemonic. Options are 'en', 'es', 'ja', 'pt', 'zh_Hans' and 'any', defaults to 'any'." />
+    <option name="passphrase" type="string" description="An optional passphrase for converting the mnemonic to a seed." />
+    <argument name="WORD" stdin="true" limit="-1" type="string" description="The set of words that that make up the mnemonic. If not specified the words are read from STDIN." />
+    <define name="BX_EC_ELECTRUM_TO_SEED_PASSPHRASE_UNSUPPORTED" value="The passphrase option requires an ICU build." />
+  </command>
+
```
Comment (evoskuil) [inline model/generate.xml]: Commands should be sorted into the generation xml alphabetically (ascii) by symbol value.

File: model/generate.xml [hunk 6]
```diff
@@ -420,6 +420,20 @@
     <define name="BX_EC_MNEMONIC_TO_SEED_INVALID_IN_LANGUAGES" value="WARNING: The specified words are not a valid mnemonic in any supported dictionary." />
   </command>
 
+  <command symbol="electrum-new" output="string" multipleX="true" category="WALLET" description="Create a mnemonic seed (Electrum) from entropy. WARNING: mnemonic should be created from properly generated entropy.">
+    <option name="prefix" type="electrum" description="The electrum seed type identifier to use. Options are 'standard', 'witness', and 'two_factor_authentication', defaults to 'standard'." />
```
Comment (evoskuil) [inline model/generate.xml]: Especially for command line (typing) explicitness should be balanced against verbosity, even in these commands where we have single letter shortcuts.  "two_factor_authentication" is a bit much. Maybe something like "dual".

**Comment by thecodefactory** [comment]
Updated

File: test/commands/electrum-to-seed.cpp (hunk 3 above)
Comment (thecodefactory) [reply test/commands/electrum-to-seed.cpp]: Any words can be decoded in electrum, although these words would not validate.

There's nothing visibly symmetric about the tests either, since the seed provided as input is an entropy seed, not the electrum seed that is decoded from the mnemonic.

**Comment by thecodefactory** [comment]
Note that a later pull request will require that we loosen the bit length requirements of bx seed command, which has a minimum bit length of 128 and requires that the value be divisible by 8.  The default electrum entropy input length is 132 bits and is padded to up to 148 bits in order to create their 13 word mnemonics.  We don't yet do any input massaging, but seed could perhaps have an electrum option to generate something compatible with their defaults.

**Comment by evoskuil** [comment]
Base16 output doesn't allow us to emit a seed that isn't byte-aligned (bits divisible by 8).

**Comment by thecodefactory** [comment]
```
diff --git a/src/commands/seed.cpp b/src/commands/seed.cpp
index bd6ec9d..bcf4af3 100644
--- a/src/commands/seed.cpp
+++ b/src/commands/seed.cpp
@@ -34,8 +34,8 @@ console_result seed::invoke(std::ostream& output, std::ostream& error)

     // These are soft requirements for security and rationality.
     // We use bit vs. byte length input as the more familiar convention.
-    if (bit_length < minimum_seed_size * byte_bits ||
-        bit_length % byte_bits != 0)
+    if (bit_length < minimum_seed_size * byte_bits)/*  || */
+        /* bit_length % byte_bits != 0) */
     {   
         error << BX_SEED_BIT_LENGTH_UNSUPPORTED << std::endl;
         return console_result::failure;
```

```
$ bx seed -b 132 | bx electrum-new
hobby fortune patch fury unusual heavy mix mass abuse distance furnace aerobic
$ bx seed -b 148 | bx electrum-new
toy plug siege worry mail cannon clog come super shoe artist soccer slide
```

The main point is that we will also later need some way to generate electrum compatible entropy so that users that want to remain compatible can easily do so.
